front/components/agent_builder/AgentBuilderContext.tsx

```diff
@@ -0,0 +1,68 @@
+import { createContext, useState } from "react";
+import { useEffect } from "react";
+
+import { mcpServerViewSortingFn } from "@app/lib/actions/mcp_helper";
+import type { MCPServerViewType } from "@app/lib/api/mcp";
+import type { AppType, DataSourceViewType, SpaceType } from "@app/types";
+
+type AgentBuilderContextType = {
+  dustApps: AppType[];
+  dataSourceViews: DataSourceViewType[];
+  spaces: SpaceType[];
+  mcpServerViews: MCPServerViewType[];
+  isPreviewPanelOpen: boolean;
+  setIsPreviewPanelOpen: React.Dispatch<React.SetStateAction<boolean>>;
+};
+
+export const AgentBuilderContext = createContext<AgentBuilderContextType>({
+  dustApps: [],
+  dataSourceViews: [],
+  spaces: [],
+  mcpServerViews: [],
+  isPreviewPanelOpen: true,
+  setIsPreviewPanelOpen: () => {},
+});
+
+export function AgentBuilderProvider({
+  dustApps,
+  dataSourceViews,
+  spaces,
+  mcpServerViews,
+  children,
+}: Omit<
+  AgentBuilderContextType,
+  "isPreviewPanelOpen" | "setIsPreviewPanelOpen"
+> & {
+  children: React.ReactNode;
+}) {
+  const [isPreviewPanelOpen, setIsPreviewPanelOpen] = useState(false);
```

Comment:
I think I know why we don't see the hydration error in assistant builder but in agent builder when we have the exact same logic. In the current Assistant builder we use `AppContentLayout`, which has [this hack](https://stackoverflow.com/questions/73162551/how-to-solve-react-hydration-error-in-nextjs). I don't think we want to use AppContentLayout here, so we should add this hack in AgentBuilderLayout?